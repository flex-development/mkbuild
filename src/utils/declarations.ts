/**
 * @file Utilities - declarations
 * @module mkbuild/utils/declarations
 */

import type { OutputFile } from 'src/interfaces'
import type {
  CompilerHost,
  CompilerOptions,
  WriteFileCallback
} from 'typescript'

/**
 * Outputs TypeScript declarations to a virtual file system.
 *
 * @async
 *
 * @param {OutputFile[]} [outputs=[]] - Output file objects
 * @return {Promise<Map<string, string>>} Virtual file system
 */
const declarations = async (
  outputs: OutputFile[] = []
): Promise<Map<string, string>> => {
  // do nothing if no outputs to operate on
  if (outputs.length === 0) return new Map()

  /**
   * Declaration output file objects.
   *
   * @const {OutputFile[]} dts
   */
  const dts: OutputFile[] = outputs.filter(output => output.declaration)

  // do nothing if no declaration file outputs to operate on
  if (dts.length === 0) return new Map()

  const { createCompilerHost, createProgram } = await import('typescript')

  /**
   * Virtual file system.
   *
   * @const {Map<string, string>} vfs
   */
  const vfs = new Map()

  /**
   * TypeScript compiler options.
   *
   * @const {CompilerOptions} compilerOptions
   */
  const compilerOptions: CompilerOptions = {
    allowJs: true,
    declaration: true,
    emitDeclarationOnly: true,
    skipLibCheck: true
  }

  /**
   * TypeScript compiler host.
   *
   * @const {CompilerHost} host
   */
  const host: CompilerHost = createCompilerHost(compilerOptions)

  /**
   * Original `host.readFile` function.
   *
   * @const {CompilerHost['readFile']} _readFile
   */
  const _readFile: CompilerHost['readFile'] = host.readFile.bind(host)

  /**
   * Reads a file from {@link vfs}.
   *
   * @param {string} filename - Name of file to write
   * @return {string | undefined} File content
   */
  const readFile = (filename: string): string | undefined => {
    return dts.find(o => o.src === filename)?.contents ?? _readFile(filename)
  }

  /**
   * Writes a file to {@link vfs}.
   *
   * @param {string} filename - Name of file to write
   * @param {string} contents - Content to write
   * @return {void} Nothing when complete
   */
  const writeFile: WriteFileCallback = (
    filename: string,
    contents: string
  ): void => void vfs.set(filename, contents)

  host.readFile = readFile
  host.writeFile = writeFile

  // emit declarations to virtual file system
  createProgram([...dts.map(o => o.src)], compilerOptions, host).emit()

  return vfs
}

export default declarations
