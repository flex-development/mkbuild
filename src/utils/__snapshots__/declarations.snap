// Vitest Snapshot v1

exports[`utils:unit/declarations > should return vfs with emitted declarations 1`] = `
Map {
  "/__fixtures__/sum-of-intervals.d.ts" => "/**
 * @file Fixtures - sumOfIntervals
 * @module fixtures/sumOfIntervals
 * @see https://codewars.com/kata/52b7ed099cdc285c300001cd
 */
/**
 * Given an array of intervals, all of varying lengths, the function returns the
 * sum of all interval lengths. **Overlapping intervals will be counted once**.
 *
 * @example
 *  sumOfIntervals([[1, 5]]) // 4
 * @example
 *  sumOfIntervals([[1, 4], [7, 10], [3, 5]]) // 7
 * @example
 *  sumOfIntervals([[1, 5], [10, 15], [-1, 3]]) // 11
 *
 * @param {[number, number][]} intervals - Intervals
 * @return {number} Sum of all interval lengths
 */
declare const sumOfIntervals: (intervals: [number, number][]) => number;
export default sumOfIntervals;
",
}
`;
